### PR TITLE
fix: fixed No deserialization found for type `Uri?` #426

### LINKED ISSUE #426: No deserialization found for type `Uri?`
I tried uploading a file to `serverpod_cloud_storage`. It works fine. But when I try getting the public url, I ran into this exception.

### Debugger traces

```console
args: {
  "path" : "PhotoViewer.m",
  "auth" : "1:P4OpgPmsZ289mQnWII0cET17c5tW0CnJ",
  "method" : "getDownloadUrl"
}

e: FormatException (FormatException: No deserialization found for type Uri?)

endpoint: "assets"

method: "getDownloadUrl"
```

### Endpoint code

```dart
import 'package:serverpod/serverpod.dart';

class AssetsEndpoint extends Endpoint {
  Future<String?> getUploadDescription(Session session, String path) async {
    try {
      session.log('Path: $path');
      return session.storage.createDirectFileUploadDescription(
        storageId: 'public',
        path: path,
      );
    } on Exception catch (e) {
      session.log(e.toString(), level: LogLevel.error);
      return null;
    }
  }

  Future<bool> verifyUpload(Session session, String path) async {
    return await session.storage.verifyDirectFileUpload(
      storageId: 'public',
      path: path,
    );
  }

  Future<Uri?> getDownloadUrl(Session session, String path) async {
    session.log('Path: $path');
    return session.storage.getPublicUrl(
      storageId: 'public',
      path: path,
    );
  }
}
```

### I use in flutter code

```dart
uploading
? const CircularProgressIndicator.adaptive()
: ElevatedButton(
    onPressed: () async {
      FilePickerResult? result =
          await FilePicker.platform.pickFiles(
        allowMultiple: false,
      );
      String? uploadDescription;
      if (result != null) {
        setState(() {
          uploading = true;
        });
        var name = result.files.single.name;
        uploadDescription = await AppConstants.client.assets
            .getUploadDescription(name);
        var bytes =
            File(result.files.single.path!).readAsBytesSync();
        bool success = false;
        if (uploadDescription != null) {
          var uploader = FileUploader(uploadDescription);
          var a = ByteData.sublistView(bytes);
          success = await uploader.uploadByteData(a);
          var uploaded =
              await AppConstants.client.assets.verifyUpload(name);
          var exists = await AppConstants.client.assets
              .checkFileExists(name);
          print('Uploaded: $uploaded');
          print('Exists: $exists');
          if (exists) {
            var a = await AppConstants.client.assets
                .getDownloadUrl(result.files.single.name);
            print(a?.toString());
            await Fluttertoast.showToast(
              msg: '$uploadDescription  $success',
              toastLength: Toast.LENGTH_SHORT,
              gravity: ToastGravity.BOTTOM,
              timeInSecForIosWeb: 5,
              backgroundColor:
                  success ? Colors.green : Colors.red,
              textColor: Colors.white,
              fontSize: 16.0,
            );
          }
          setState(() {
            uploading = false;
          });
        }
      }
    },
    child: const Text('Upload and get URL'),
  ),
```

### serverpod_cloud_storage table 

| id      | storageId    | path    | addedTime    | expiration    | byteData    | verified    |
| :-----: | :---: | :---: | :---: | :---: | :---: | :---: |
| 1 | public   | PhotoViewer.m   | 2022-12-03 21:39:10.476641   | NULL   | BINARY (13 KB)   | TRUE   |

### PR DESCRIPTION
**Pull request by yahu1031**
Fixed #426

## Pre-launch Checklist

- [X] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [X] This update cointains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [X] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [X] All existing and new tests are passing.
- [X] Any breaking changes are documented below.

## Breaking changes

No breaking changes.

### CODE REVIEW

**Review by vlidholt — CHANGES_REQUESTED** [review]
Please refer to https://github.com/serverpod/serverpod/pull/493 to see what is required to add a new type to serialization.

Specifically:
1. There must be tests.
2. There must be a strategy to do serialization to the database.

File: tests/serverpod_test_server/generated/tables.pgsql [hunk 1]
```diff
@@ -84,7 +97,8 @@ CREATE TABLE "types" (
   "aDouble" double precision,
   "aDateTime" timestamp without time zone,
   "aString" text,
-  "aByteData" bytea
+  "aByteData" bytea,
+  "anUri" character varying
```
Comment (vlidholt) [inline tests/serverpod_test_server/generated/tables.pgsql:101]: This should be of type `text` to be consistent with other database fields. (`text` and `varchar` are interchangeable in Postgres)

**Review by vlidholt — CHANGES_REQUESTED** [review]

**Comment by vlidholt** [comment]
@yahu1031 There is no test where the Uri is passed to the server or written to the database. That needs to be there.

File: tests/serverpod_test_server/generated/tables.pgsql (hunk 1 above)
Comment (yahu1031) [reply tests/serverpod_test_server/generated/tables.pgsql:101]: Okk i will change it. Please LMK anything else.

**Comment by yahu1031** [comment]
Yes that was pending, I'm not aware of how to implement. Will check it.

**Comment by fischerscode** [comment]
@yahu1031 Are you still working on this? In case you want, I could also provide a PR.

#641 Might be useful for reference, but some parts (database representation is a String) don't have to be done in order to support `Uri`s.

**Comment by yahu1031** [comment]
> @yahu1031 Are you still working on this? In case you want, I could also provide a PR.
> 
> #641 Might be useful for reference, but some parts (database representation is a String) don't have to be done in order to support `Uri`s.

I'm holding this back cause, that is a very super specific case. Unless you want a Uri. But we ideally don't have such a use-case cause when we `myUri.toString()` it, we get the Url. And also when we work on API endpoints, We can't get `Uri` as result to communicate with other languages. Other than my report, I don't see this issue from any end. LMK what do you think? If you would like me to finish it off, please help me out by just supporting me to write the test cases, I'm a little bit confused.

**Comment by fischerscode** [comment]
I don't need it in my project. I just remembered, there was one more pending type addition. I think there should be types with build in support and types with documentation support.

A type with build in support should work out of the box, with no additional work required by the developer. A type with documentation support should not be should not work out of the box, but there should be documentation how to use this type in a project using `extraClasses` or `customConstructors`. For `Uri` however, we'd have to add something like `customSerializers`, I think. One might argue, [UuidValue](https://github.com/serverpod/serverpod/pull/641) should be a 'documented type', too.

---

Regarding tests, I think it would be 'enough' mirror this lines with `Uri`:
- [x] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/lib/src/protocol/types.yaml#L11
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/lib/src/protocol/nullability.yaml#L24-L25
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/lib/src/protocol/nullability.yaml#L55-L58
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/lib/src/protocol/object_with_maps.yaml#L9
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/lib/src/protocol/object_with_maps.yaml#L16
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/lib/src/endpoints/basic_types.dart#L33-L35
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/connection_test.dart#L818-L826
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/connection_test.dart#L843
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/connection_test.dart#L869
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/connection_test.dart#L899
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L37
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L49
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L63-L64
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L134-L142
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L189-L192
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L202
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L211
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L274-L291
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L304-L305
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L345-L349
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L356-L359
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L405-L412
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L439-L442
- [ ] https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tests/serverpod_test_server/test/serialization_test.dart#L466-L468

File: tests/serverpod_test_server/lib/src/generated/object_with_uri.dart [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+typedef ObjectWithUriExpressionBuilder = _i1.Expression Function(
+    ObjectWithUriTable);
+
+class ObjectWithUriTable extends _i1.Table {
+  ObjectWithUriTable() : super(tableName: 'object_with_uri');
+
+  final id = _i1.ColumnInt('id');
+
+  final url = _i1.ColumnSerializable('url');
```
Comment (fischerscode) [inline tests/serverpod_test_server/lib/src/generated/object_with_uri.dart:191]: I think a `ColumnString` would be better.

Change would go here.
https://github.com/serverpod/serverpod/blob/046b95537404f340784fbef9a15d1ca62b6e332f/tools/serverpod_cli/bin/generator/types.dart#L173-L186

**Comment by vlidholt** [comment]
This would be included with the above mentioned things fixed and most importantly, tests in place.

**Comment by yahu1031** [comment]
Will use this PR as a ref and recreate a new PR with resolved conflicts. Closing this PR for time being.

**Pull request closed without merging**